### ISSUE TITLE
shutdown event not handled by worker-runner and docker-worker

### ISSUE THREAD

**Issue by lotas** [taskcluster/taskcluster#6314]
reported by @jcristau 

https://paste.debian.net/hidden/e5fc183d/ is from a worker's logs, namely https://firefox-ci-tc.services.mozilla.com/provisioners/gecko-3/worker-types/decision-gcp/workers/us-west1/4821576692450676629?sortBy=started&sortDirection=desc
it says `"Power key pressed"`, part way through https://firefox-ci-tc.services.mozilla.com/tasks/UQNbiVwhRt6Setm7s6Uudg/runs/0, and apparently ~immediately stops running that task, but doesn't report anything so instead of ending with worker-shutdown it waits until claim-expired.
but then it looks like it immediately claims https://firefox-ci-tc.services.mozilla.com/tasks/JOH-4X6nTVqF-74kOC0GSw/runs/0, just before it actually shuts down, so that also ends up claim-expired.
is this a (known?) worker issue, or something that should/could be handled on the services side?

This is a GCP instance with worker-runner and docker-worker v38.0.5
```
Jun 06 09:20:00Z gecko-3-decision-gcp-ygbwfquxssmrk6-unlmu7w systemd-logind Power key pressed.
Jun 06 09:20:00Z gecko-3-decision-gcp-ygbwfquxssmrk6-unlmu7w systemd-logind Powering Off...
```
No reaction on this event, and worker keeps polling for new tasks

**Comment by djmitche** [comment]
It's hard to tell without the full logs, but a simple explanation of what happened is:

* The worker was executing [the backfill task](https://firefox-ci-tc.services.mozilla.com/tasks/UQNbiVwhRt6Setm7s6Uudg/runs/0)
* 9:20:00 - logind got the "Power button pressed" signal from ACPI and initiated a shutdown, but ultimately didn't manage to tell the worker about that (I don't know what signal it should send to what process?)
* 9:20:00 - docker-worker continued executing the backfill task
* 9:20:02 - the backfill task entered another phase of its execution and stopped calling createTask
* 9:20:27 - [A second task](https://firefox-ci-tc.services.mozilla.com/tasks/JOH-4X6nTVqF-74kOC0GSw) was returned from claimWork (this worker has capacity 2, so this doesn't mean it stopped executing the first task)
* (a little later) either logind finally managed to kill the worker process or the entire instance halted

That a new task arrived within 25 seconds of the shutdown signal doesn't seem too surprising.

It's probably worth looking up what logind does when it gets that "Power key pressed" event. At a guess, it sends some signal to every process, then waits 30 seconds, then halts the system. Maybe docker-worker didn't receive that signal or maybe it was not the signal docker-worker was expecting?

Docker-worker's ["shutdown manager"](https://github.com/taskcluster/taskcluster/blob/main/workers/docker-worker/src/shutdown_manager.js) expects a SIGTERM. It logs immediately when it gets that signal, and that message isn't in the pastebin. To my knowledge, worker-runner does not specifically handle this sort of signal, but the default handling of SIGTERM is to terminate the process, so it's likely that worker-runner would just disappear when this occurs.

tl;dr: more information needed :)

**Comment by djmitche** [comment]
What caused the power-down? Was this a spot termination?

**Comment by lotas** [comment]
>What caused the power-down? Was this a spot termination?
is there a way to be sure? There's nothing more than that in the papertrail logs. 
In services logs the instance has state "TERMINATED" during worker scanner run later

**Comment by lotas** [comment]
ah, got it! No, worker manager didn't shut it down. 

Regarding `claim-expired`:
<img width="1874" alt="image" src="https://github.com/taskcluster/taskcluster/assets/83861/73a825ba-d6e9-4024-9117-a3a2d4ad1aa0">

it happens, but comparing to the total number of tasks we run, this is <0.5%. 
Most of those are coming from worker pools that run docker-worker ~v29, <v38, so it's possible that some issues were already solved since then? 

I don't think it is realistic to get rid of all the race conditions. There is a possibility to minimise the impact by playing with `takenUntil` values (so the tasks could be picked up by different workers faster), but this would lead to higher load.

**Comment by djmitche** [comment]
0.5% seems a small number to worry about, indeed.

Getting the shutdown_manager to actually recognize the SIGTERM would at least turn those into `worker-shutdown`, which would help eliminate these cases from the metrics, but that's probably a lot of investigative work for little benefit.

**Comment by Eijebong** [comment]
Docker-worker isn't a thing anymore since #8437
Closing
